### PR TITLE
#407: FtSecure (for https)

### PR DESCRIPTION
**Pull request by bdragan**
#407: FtSecure (for https).

### CODE REVIEW

**Comment by davvd** [comment]
@bdragan Thanks for your pull request, let me find someone who can review it

**Comment by davvd** [comment]
@pinaf please review, thanks

File: src/main/java/org/takes/http/FtSecure.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    private final transient Front front;
+
+    /**
+     * Ctor.
+     * @param tks Take
+     * @throws IOException If fails
+     */
+    public FtSecure(final Take tks) throws IOException {
+        // @checkstyle MagicNumber (1 line)
+        this(new BkBasic(tks), 443);
```
Comment (pinaf) [inline src/main/java/org/takes/http/FtSecure.java]: @bdragan here let's do `this(tks, 443)`, leaving the creation of the back to the other constructor.

File: src/main/java/org/takes/http/FtSecure.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            SSLServerSocketFactory.getDefault().createServerSocket(port)
+        );
+    }
+
+    /**
+     * Ctor.
+     * @param bck Back
+     * @param skt Server socket
+     */
+    public FtSecure(final Back bck, final ServerSocket skt) {
```
Comment (pinaf) [inline src/main/java/org/takes/http/FtSecure.java]: @bdragan do we really need this constructor to be public?

**Comment by pinaf** [comment]
@bdragan 2 comments above

**Comment by pinaf** [comment]
@bdragan also, we need to test this.

**Comment by bdragan** [comment]
@pinaf Good points, please take a look at the new commit.

File: src/test/java/org/takes/http/FtSecureTest.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        System.setProperty("javax.net.ssl.trustStorePassword", password);
+    }
+
+    /**
+     * FtSecure can work.
+     * @throws Exception If some problem inside
+     */
+    @Test
+    public void justWorks() throws Exception {
+        FtRemote.secure(new TkFork(new FkRegex("/", "hello, world!"))).exec(
```
Comment (pinaf) [inline src/test/java/org/takes/http/FtSecureTest.java]: @bdragan this class doesn't seem to be testing `FtSecure`.

File: pom.xml [hunk 4]
```diff
@@ -217,6 +217,22 @@
         </dependency>
     </dependencies>
     <build>
+		<testResources>
+        	<testResource>
```
Comment (pinaf) [inline pom.xml]: @bdragan these 2 `testResource` groups look similar to each other. Could you please explain the logic for them?

File: src/main/java/org/takes/http/FtRemote.java [hunk 5]
```diff
@@ -148,6 +173,23 @@ public boolean ready() {
     }
 
     /**
+     * Creates an instance of secure Front.
+     *
+     * @param take Take
+     * @return Secure Front
+     * @throws IOException If some problem inside
+     */
+    public static FtRemote secure(final Take take) throws IOException {
```
Comment (pinaf) [inline src/main/java/org/takes/http/FtRemote.java]: @bdragan let's move this method to the test class below.

**Comment by pinaf** [comment]
@bdragan 3 more

File: pom.xml (hunk 4 above)
Comment (bdragan) [inline pom.xml]: @pinaf This is how maven resources plugin works; I also find it a bit counter-intuitive. The first one copies all the files except the `keystore` to the target directory replacing all the property placeholders (`filtering=true`). The second one copies only the `keystore` file without processing its contents (`filtering=false`).

An example in the maven official docs can be found [here](http://maven.apache.org/plugins/maven-resources-plugin/examples/filter.html).

**Comment by bdragan** [comment]
@pinaf Done; please see the new commit.

File: src/test/java/org/takes/http/FtSecureTest.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+                        .as(RestResponse.class)
+                        .assertStatus(HttpURLConnection.HTTP_OK)
+                        .assertBody(Matchers.startsWith("hello"));
+                }
+            }
+        );
+    }
+
+    /**
+     * FtSecure can work with a broken back.
```
Comment (pinaf) [inline src/test/java/org/takes/http/FtSecureTest.java]: @bdragan let's maintain consistency with the method name. `can gracefully handle a broken Back.`

File: src/test/java/org/takes/http/FtSecureTest.java [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+     * @throws Exception If some problem inside
+     */
+    @Test
+    public void gracefullyHandlesBrokenBack() throws Exception {
+        secure(new TkFailure("Jeffrey Lebowski")).exec(
```
Comment (pinaf) [inline src/test/java/org/takes/http/FtSecureTest.java]: @bdragan `FtSecureTest.secure(`

File: src/test/java/org/takes/http/FtSecureTest.java [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        System.setProperty("javax.net.ssl.trustStorePassword", password);
+    }
+
+    /**
+     * FtSecure can work.
+     * @throws Exception If some problem inside
+     */
+    @Test
+    public void justWorks() throws Exception {
+        secure(new TkFork(new FkRegex("/", "hello, world!"))).exec(
```
Comment (pinaf) [inline src/test/java/org/takes/http/FtSecureTest.java]: @bdragan `FtSecureTest.secure(`

File: src/test/java/org/takes/http/FtSecureTest.java [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+            @Override
+            public Response act(final Request request) throws IOException {
+                MatcherAssert.assertThat(
+                    new RqPrint(request).printBody(),
+                    Matchers.containsString("Jeff")
+                );
+                return new RsText("works!");
+            }
+        };
+        secure(take).exec(
```
Comment (pinaf) [inline src/test/java/org/takes/http/FtSecureTest.java]: @bdragan `FtSecureTest.secure(`

File: src/test/java/org/takes/http/FtSecureTest.java [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+            @Override
+            public Response act(final Request req) throws IOException {
+                return new RsText(
+                    IOUtils.toString(
+                        new RqLengthAware(req).body()
+                    )
+                );
+            }
+        };
+        secure(take).exec(
```
Comment (pinaf) [inline src/test/java/org/takes/http/FtSecureTest.java]: @bdragan `FtSecureTest.secure(`

File: src/test/java/org/takes/http/FtSecureTest.java (hunk 8 above)
Comment (pinaf) [inline src/test/java/org/takes/http/FtSecureTest.java]: @bdragan do we really need a `TkFork` here?

**Comment by pinaf** [comment]
@bdragan thanks

**Comment by pinaf** [comment]
@bdragan a few more

File: src/test/java/org/takes/http/FtSecureTest.java (hunk 7 above)
Comment (bdragan) [inline src/test/java/org/takes/http/FtSecureTest.java]: @pinaf Could you please elaborate on this? A link to the relevant docs or project conventions would be sufficient.

File: src/test/java/org/takes/http/FtSecureTest.java (hunk 8 above)
Comment (bdragan) [inline src/test/java/org/takes/http/FtSecureTest.java]: Yes, [monkey see, monkey copy-paste](https://en.wikipedia.org/wiki/Monkey_see,_monkey_do). I copy-pasted the existing test here. :)

**Comment by bdragan** [comment]
@pinaf Thanks, the new commit is ready for review.

File: src/test/java/org/takes/http/FtSecureTest.java (hunk 7 above)
Comment (pinaf) [inline src/test/java/org/takes/http/FtSecureTest.java]: @bdragan We always add the full path to static method calls - not sure why qulice hasn't complained here.

File: src/test/java/org/takes/http/FtSecureTest.java (hunk 8 above)
Comment (pinaf) [inline src/test/java/org/takes/http/FtSecureTest.java]: @bdragan np. could you please edit the previous comment and address it to me? otherwhite QA will complain.

**Comment by pinaf** [comment]
@rultor merge

**Comment by pinaf** [comment]
@bdragan thanks!

File: src/test/java/org/takes/http/FtSecureTest.java [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    /**
+     * Creates an instance of secure Front.
+     *
+     * @param take Take
+     * @return Secure Front
+     * @throws IOException If some problem inside
+     */
+    public static FtRemote secure(final Take take) throws IOException {
```
Comment (yegor256) [inline src/test/java/org/takes/http/FtSecureTest.java]: @bdragan why public?

**Comment by yegor256** [comment]
@bdragan good job, one small comment above

**Comment by bdragan** [comment]
@yegor256 Thanks, good catch about the access modifier. Fixed.

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by davvd** [comment]
@elenavolokhova review this ticket please, for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor deploy

**Comment by elenavolokhova** [comment]
@davvd Looks good!

**Comment by davvd** [comment]
> @davvd Looks good!

@elenavolokhova thanks a lot

**Comment by davvd** [comment]
@pinaf added 10 mins to @elenavolokhova, for QA review, payment code is `69110532`... Thanks for your contribution, **35 mins** was added to your account, payment ID is `AP-074306799V548482X`, 114 hours and 19 mins spent total... you're getting extra minutes here (c=20)... +35 added to your rating, current score is: [+4060](http://www.netbout.com/b/35128?open=rating)

**Pull request merged**